ledger-core/virtual/handlers/handle_meta.go

```diff
@@ -33,78 +35,79 @@ type errNoHandler struct {
-		return obj.AsOf, func(ctx smachine.ConstructionContext) smachine.StateMachine {
-			ctx.SetContext(goCtx)
-			ctx.SetTracerID(traceID)
-			return &SMVStateRequest{Meta: payloadMeta, Payload: obj}
-		}
-	case *payload.VStateReport:
-		return payloadMeta.Pulse, func(ctx smachine.ConstructionContext) smachine.StateMachine {
-			ctx.SetContext(goCtx)
-			ctx.SetTracerID(traceID)
-			return &SMVStateReport{Meta: payloadMeta, Payload: obj}
-		}
-	case *payload.VDelegatedRequestFinished:
-		return payloadMeta.Pulse, func(ctx smachine.ConstructionContext) smachine.StateMachine {
-			ctx.SetContext(goCtx)
-			ctx.SetTracerID(traceID)
-			return &SMVDelegatedRequestFinished{Meta: payloadMeta, Payload: obj}
-		}
-	case *payload.VDelegatedCallRequest:
-		return payloadMeta.Pulse, func(ctx smachine.ConstructionContext) smachine.StateMachine {
-			ctx.SetContext(goCtx)
-			ctx.SetTracerID(traceID)
-			return &SMVDelegatedCallRequest{Meta: payloadMeta, Payload: obj}
+	mustReject, err := authService.IsMessageFromVirtualLegitimate(goCtx, payloadObj, payloadMeta.Sender, pr)
+	if err != nil {
+		logger.Warn(throw.W(err, "illegitimate msg", struct {
+			messageTypeID uint64
+			messageType   reflect.Type
+		}{messageTypeID: payloadTypeID, messageType: payloadType}))
+
+		return pulse.Unknown, nil, nil
+	}
+
+	if mustReject {
+		// when this flag is set, then the relevant SM has to stop asap and send negative answer
+		return pulse.Unknown, nil, throw.NotImplemented()
+	}
+
+	if pn, sm := func () (pulse.Number, smachine.StateMachine) {
+		switch obj := payloadObj.(type) {
+		case *payload.VCallRequest:
+			return currentPulse, &SMVCallRequest{Meta: payloadMeta, Payload: obj}
+		case *payload.VCallResult:
+			return currentPulse, &SMVCallResult{Meta: payloadMeta, Payload: obj}
+		case *payload.VStateRequest:
+			return obj.AsOf, &SMVStateRequest{Meta: payloadMeta, Payload: obj}
+		case *payload.VStateReport:
+			return currentPulse, &SMVStateReport{Meta: payloadMeta, Payload: obj}
+		case *payload.VDelegatedRequestFinished:
+			return currentPulse, &SMVDelegatedRequestFinished{Meta: payloadMeta, Payload: obj}
+		case *payload.VDelegatedCallRequest:
+			return currentPulse, &SMVDelegatedCallRequest{Meta: payloadMeta, Payload: obj}
+		case *payload.VDelegatedCallResponse:
+			return currentPulse, &SMVDelegatedCallResponse{Meta: payloadMeta, Payload: obj}
+		default:
+			return 0, nil
 		}
-	case *payload.VDelegatedCallResponse:
-		return payloadMeta.Pulse, func(ctx smachine.ConstructionContext) smachine.StateMachine {
+	}(); sm != nil {
+		return pn, func(ctx smachine.ConstructionContext) smachine.StateMachine {
 			ctx.SetContext(goCtx)
 			ctx.SetTracerID(traceID)
-			return &SMVDelegatedCallResponse{Meta: payloadMeta, Payload: obj}
-		}
-	default:
-		panic(errNoHandler{
-			messageTypeID: payloadTypeID,
-			messageType:   payloadType,
-		})
+			return sm
+		}, nil
 	}
+
+	logger.Warn(errNoHandler{messageTypeID: payloadTypeID, messageType: payloadType})
```

Comment:
how is this different from `default:` in switch?
It looks the same to me.
You can just ommit default or move this there